staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go

```diff
@@ -0,0 +1,169 @@
+}
+
+// OpenAPIAggregationController periodically check for changes in OpenAPI specs of APIServices and update/remove
+// them if necessary.
+type OpenAPIAggregationController struct {
+	openAPIAggregationManager OpenAPIAggregationManager
+	queue                     workqueue.RateLimitingInterface
+	downloader                *openAPIDownloader
+
+	// To allow injection for testing.
+	syncHandler func(key string) (syncAction, error)
+}
+
+func NewOpenAPIAggregationController(downloader *openAPIDownloader, openAPIAggregationManager OpenAPIAggregationManager) *OpenAPIAggregationController {
+	c := &OpenAPIAggregationController{
+		openAPIAggregationManager: openAPIAggregationManager,
+		queue: workqueue.NewNamedRateLimitingQueue(
+			workqueue.NewItemExponentialFailureRateLimiter(successfulUpdateDelay, failedUpdateMaxExpDelay), "APIServiceOpenAPIAggregationControllerQueue1"),
+		downloader: downloader,
+	}
+
+	c.syncHandler = c.sync
+
+	return c
+}
+
+func (c *OpenAPIAggregationController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	glog.Infof("Starting OpenAPIAggregationController")
+	defer glog.Infof("Shutting down OpenAPIAggregationController")
+
+	go wait.Until(c.runWorker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (c *OpenAPIAggregationController) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (c *OpenAPIAggregationController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	defer c.queue.Done(key)
+	if quit {
+		return false
+	}
+
+	action, err := c.syncHandler(key.(string))
+	if err == nil {
+		c.queue.Forget(key)
+	} else {
+		utilruntime.HandleError(fmt.Errorf("loading OpenAPI spec for %q failed with: %v", key, err))
+	}
+
+	switch action {
+	case syncRequeue:
+		c.queue.AddAfter(key, successfulUpdateDelay)
+	case syncRequeueRateLimited:
+		c.queue.AddRateLimited(key)
+	case syncNothing:
+	}
+
+	return true
+}
+
+func (c *OpenAPIAggregationController) sync(key string) (syncAction, error) {
+	handler, etag, exists := c.openAPIAggregationManager.GetApiServiceInfo(key)
+	if !exists || handler == nil {
+		return syncNothing, nil
+	}
+	returnSpec, newEtag, httpStatus, err := c.downloader.Download(handler, etag)
+	switch {
+	case err != nil:
+		return syncRequeueRateLimited, err
+	case httpStatus == http.StatusNotFound || returnSpec == nil:
+		return syncRequeueRateLimited, fmt.Errorf("OpenAPI spec does not exists")
+	case httpStatus == http.StatusOK:
+		if err := c.openAPIAggregationManager.UpdateApiServiceSpec(key, returnSpec, newEtag); err != nil {
+			return syncRequeueRateLimited, err
+		}
+	}
+	return syncRequeue, nil
+}
+
+func (c *OpenAPIAggregationController) AddAPIService(handler http.Handler, apiService *apiregistration.APIService) {
+	if err := c.openAPIAggregationManager.AddUpdateApiService(handler, apiService); err != nil {
```

Comment:
@mbohlool in @frobware's tests we see download errors for the local api groups after some minutes. Can it be that we have to filter out the local services from the controller logic?

Reply:
This plus squashing blocks my lgtm.

Reply:
This should be fixed with missing case for unmodified files. The squashing however is not that easy The delegation chain commit is in between OpenAPI commits. I had many commits and tried my best to reduce number of commits and make them as self contain and descriptive as possible. Let see how do you like that. Will push and ping you when it is ready.

Reply:
I was running @frobware's test and noticed what I suspected above: all local APIServices cause a download of the local spec. This takes roughly one second per APIService (i.e. per GroupVersion) on my machine, i.e. I see downloads and a busy CPU for rougly 20 seconds. After this the etag is recorded and the download is quick. But after each download the filtered spec is stored in the aggregator spec map. So you have each GVK spec twice in there: once from the delegation chain, once from the download. This slows down the merging by at least a factor of two (if don't have quadratic runtime somewhere, then it much worse). And if one of the spec changes over time, i.e. from CRDs (if we support them in OpenAPI), you have two name conflicts and renaming in the merge algorithm.

As before, I suggest to stop downloading local specs for now.